colossalai/zero/sharded_optim/sharded_optim_v3.py

```diff
@@ -0,0 +1,155 @@
+                 mp_process_group: Optional[ProcessGroup] = None) -> None:
+        super().__init__(optimizer)
+        self.model: Union[nn.Module, ShardedModelV2] = sharded_model
+        self.model_is_sharded = isinstance(sharded_model, ShardedModelV2)
+        self.device = torch.cuda.current_device() if not cpu_offload else torch.device('cpu')
+        self.optim_state: OptimState = OptimState.UNSCALED
+        self.dp_process_group = dp_process_group or gpc.get_group(ParallelMode.DATA)
+        self.mp_process_group = mp_process_group or gpc.get_group(ParallelMode.MODEL)
+        self.shard_strategy = TensorShardStrategy(dp_process_group)
+        # Grad scaler
+        self.grad_scaler = DynamicGradScaler(initial_scale=initial_scale,
+                                             min_scale=min_scale,
+                                             growth_factor=growth_factor,
+                                             backoff_factor=backoff_factor,
+                                             growth_interval=growth_interval,
+                                             hysteresis=hysteresis,
+                                             max_scale=max_scale)
+        self._found_overflow: Tensor = torch.FloatTensor([0]).to(self.device)
+
+        # Store fp32 params
+        self.master_params: Dict[Parameter, Union[Tensor, ShardedTensor]] = {}
+
+        for group in optimizer.param_groups:
+            for p in group['params']:
+                fp32_param: Tensor = p.ca_attr.data if hasattr(p, 'ca_attr') else p.data
+                if torch.is_floating_point(fp32_param) and fp32_param.dtype != torch.float:
+                    fp32_param = fp32_param.float()
+                if hasattr(p, 'ca_attr'):
+                    assert p.ca_attr.is_sharded, 'ShardedAdam can be only used with sharded model'
+                    self.master_params[p] = fp32_param
+                else:
+                    self.master_params[p] = ShardedTensor(fp32_param, process_group=dp_process_group)
+
+    def step(self, *args, **kwargs):
+        # unscale grads if scaled
+        if self.optim_state == OptimState.SCALED:
+            self._unscale_grads()
+
+        found_inf = self._check_overflow()
+        self.grad_scaler.update(found_inf)
+
+        if found_inf:
+            self.zero_grad()
+            return
+
+        # Write master param to p.data
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                if hasattr(p, 'ca_attr'):
+                    p.data = self.master_params[p]
+                else:
+                    self.shard_strategy.shard([self.master_params[p]])
+                    p.data = self.master_params[p].payload
+        ret = self.optim.step(*args, **kwargs)
+        # Write master param to payload
+        for group in self.optim.param_groups:
+            for p in group['params']:
+                if hasattr(p, 'ca_attr'):
+                    p.ca_attr.data = p.data
```

Comment:
Did you make sure the param copy is fp32 -> fp16?

Reply:
If it is ZeRO3, it should copy cpu fp32 -> gpu fp16.

Reply:
the setter of ShardedParam data is not good, you can add a new function as copy_data(self, t)?

Reply:
OK